### PR TITLE
WIP: Mission sync with master

### PR DESCRIPTION
**Pull request by AndreasAntener**
**needs a rebase and testing**
- weathervane on takeoff
- separate cruising speeds for VTOL in MC and FW
- cruising speed resets
- allow to update cruising speed at any time during a mission
- mission work item logic is more clear
- fixed double execution of certain work item states
- moved VTOL transition target position generation to mission code and set always

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Thanks!

**Comment by LorenzMeier** [comment]
What is the testing remaining / required here?

File: src/modules/navigator/navigator_main.cpp [hunk 1]
```diff
@@ -729,26 +747,52 @@ Navigator::get_altitude_acceptance_radius()
...
+	_mission_cruising_speed_mc = -1.0f;
+	_mission_cruising_speed_fw = -1.0f;
+}
+
 float
 Navigator::get_cruising_throttle()
 {
 	/* Return the mission-requested cruise speed, or default FW_THR_CRUISE value */
-	if (_mission_throttle > 0.0f) {
+	if (_mission_throttle > FLT_EPSILON) {
```
Comment (AndreasAntener) [inline src/modules/navigator/navigator_main.cpp]: note to self, unnecessary change

**Comment by AndreasAntener** [comment]
A rebase and (at least) thorough SITL mission testing, especially all VTOL items, with cruising speed and fixed throttle operation.

**Comment by dagar** [comment]
Is the scenario for updating cruise speed receiving MAV_CMD_DO_CHANGE_SPEED? What if that was handled directly in the fw/mc position controllers?

**Comment by AndreasAntener** [comment]
I don't see why they should handle that themselves, the setpoint interface is used for both and the handling can therefore be done in one place.

**Comment by dagar** [comment]
I was thinking LOITER or future use in ALTCTL/POSCTL if we made the stick
input more configurable.

**Comment by LorenzMeier** [comment]
@AndreasAntener Would you mind a rebase? Sorry for the ripples in master, but they do benefit everybody. If we stop doing cleanups the whole system will just bit-rot.

**Comment by dagar** [comment]
Let's lock down navigator code style once this is merged.

File: src/modules/navigator/mission.cpp [hunk 2]
```diff
@@ -567,15 +599,19 @@ Mission::set_mission_items()
 		/* move to land wp as fixed wing */
 		if (_mission_item.nav_cmd == NAV_CMD_VTOL_LAND
```
Comment (dagar) [inline src/modules/navigator/mission.cpp]: Doesn't this case overlap with "move to landing waypoint before descent if necessary" below?

**Comment by AndreasAntener** [comment]
@dagar that would be good ;)
Regarding your question from #5808, this branch handles interrupting non-default work items properly.

File: src/modules/navigator/mission.cpp (hunk 2 above)
Comment (AndreasAntener) [inline src/modules/navigator/mission.cpp]: It takes precedence over the one below, and the one below kicks in too in the end. The sequence is long (to make sure no waypoint gets missed):
- vtol land
- move to land (as FW)
- transition
- move to land (as MC)

**Comment by dagar** [comment]
Reviewed and partially tested. I'll try to thoroughly test everything new once it's rebased.

I won't try to suggest doing something about it right now, but I find the work item handling in mission to be very difficult to follow. I probably could have extended it to handle FW loiter while climbing/descending to a waypoint, but that ended up crammed into is_mission_item_reached(). Maybe a switch statement for work_item_type and breaking out more functionality into functions would help make it easier to understand and extend.

**Comment by AndreasAntener** [comment]
@dagar thanks. I have to admit that one of my initial goals was something like this: https://github.com/PX4/Firmware/blob/master/src/modules/navigator/rtl.cpp#L135
That doesn't really change the complexity of "expanded" mission items, but maybe increases readability. It would separate state transitions from actual functionality.

**Comment by dagar** [comment]
Yes, that's exactly what I was thinking.

**Comment by LorenzMeier** [comment]
@AndreasAntener If you rebase this I'm happy to merge it.

**Comment by AndreasAntener** [comment]
No worries, need to get back to it. @julianoes did you want to pick some parts over now for #5993?

**Comment by LorenzMeier** [comment]
@AndreasAntener Any chance of rebasing today?

**Comment by LorenzMeier** [comment]
Rebased and force-pushed.

**Comment by LorenzMeier** [comment]
Logic tested in flight (bad tracking due to poor GPS, but the mission itself was performed correctly): http://review.px4.io/plot_app?log=43972e0a-a042-4c94-b81f-0197dcbcd71b
